src/NHibernate/Criterion/SQLCriterion.cs

```diff
@@ -59,5 +62,19 @@ public override string ToString()
 		{
 			return _sql.ToString();
 		}
+
+		/// <summary>
+		/// Provide list of criteria aliases that's used in SQL projection.
+		/// To be replaced with SQL aliases.
+		/// </summary>
+		public SQLCriterion AddAliases(params string[] criteriaAliases)
```

Comment:
After a close inspection of the changes I concluded that this method is unnecessary. All information has already presented in the query.

We have 2 ways forward:
- Extract aliases from SQLCriterion/SQLProjection with Regex; This approach requires just small adjustments, but I don't quite like it as it requires regex processing, which could be expensive. This can be slightly optimized if aliases are extracted on SQLCriterion/SQLProjection is construction.
- Get aliases from Criteria.

Reply:
> Extract aliases from SQLCriterion/SQLProjection with Regex

I don't like any automatic parsing routines and intentionally made it manual. All automatic parsing logic should at least be optional - as in most cases it's not required. So I don't like your PR which now adds mandatory Replace calls for all criteria aliases in query.

And initially I wanted to support not only aliases but any projections in this template:
```C#
Projections.SQL(" {value1} + {x}.Value2")
.AddAliases("x")
.AddProjection("value1", Projections.Property("Value1"))
```

But projections support requires `SqlString.Replace(sqlString, sqlString)` which is missing. And I left implementation for later.

So I'm for manual explicit alias registration (similar how it's done in hql SQL - `ISQLQuery.AddEntity`). As convenience we can add method (like `ResolveAliases`) which would parse string via regex and populate `_criteriaAliases` but don't see necessity in it (user can implement it himself).

> Get aliases from Criteria.

`ICriteriaQuery` contains this data so why can't it be exposed? It feels just like proper place for retrieving such information. And I'm not sure I understand how it can be retrieved from `ICriteria`. Could you clarify?

Reply:
>And in some cases it's harmful (like when user wants `{alias}` be present in SQL projection without any transformation)
> So I'm for manual explicit alias registration
> PR which now adds mandatory Replace calls for all criteria aliases in query.

I don't understand why would anyone want to leave untranslated `{something}` in the SQL. So, I don't see a way when there is an alias in SQLCriterion/SQLProjection which would not be registered in `AddAliases(...)` call.

For me it feels more natural and convenient to translate all aliases rather than manually register them for translation.

If you are talking about performance here, it is negligible in my view as CriteriaTranslator already does some unnecessary heavy lifting "just in case".

> I'm for manual explicit alias registration (similar how it's done in hql SQL - `ISQLQuery.AddEntity`)

`AddEntity` is a little different than this. It add a necessary information to the query which it would not have otherwise (entity type for example). On the other hand, aliases are already known and readily available to use, so `AddAliases` not add real value.

> > Get aliases from Criteria.

> `ICriteriaQuery `contains this data so why can't it be exposed? It feels just like proper place for retrieving such information. 

This is exactly what I meant and how I implemented it. 

> And I'm not sure I understand how it can be retrieved from `ICriteria`. Could you clarify?

I did not say `ICriteria`, I meant somewhere in the Criteria. There are aliases in `ICriteria`, but they would need to be passed to the `ICriteriaQuery` to get SQLAliases. I tried to implement it this way, but was more complex.

> And initially I wanted to support not only aliases but any projections in this template

Nothing in your PR or my addition stops you to implement this, and the solution would be orthogonal to the resolving aliases. The only similarity would be that it would be happening at the same place.

Reply:
I think we would need to invoke @fredericDelaporte so he can express his view and resolve the arguments.

Reply:
> I don't understand why would anyone want to leave untranslated {something} 

Because it's part of some unrelated string in this projection. 

> Nothing in your PR or my addition stops you to implement this

Mostly user would use either projection or criteria alias but not both - so unnecessary additional work. And it could clash with user aliases for projections. With explicit registering it's less probable.

> In the other hand, aliases are already known and readily available to use, so AddAliases not add real value

Explicity is added value for me.

> If you are talking about performance here, ... CriteriaQueryTranslator already does some unnecessary heavy lifting "just in case"

That's a not good point for adding additionally unnecessary heavy lifting. We should minimize it instead. 
And yes that's my main concern - I really don't like that for each SQL projection/criterion we generate dictionary with existing aliases, then generate possible aliases to replace in template and call Replace for them.

In any way your approach is different implementation so if @fredericDelaporte also think it's better just create own PR and I will close mine.

Reply:
>it's better just create own PR and I will close mine.

Created

Reply:
As I am mostly a HQL and Linq "guy", having not used a lot Criteria, I do not think my view should be taken with a lot of weight.

This said, I am more in favor of Alexander viewpoint, at least about its rationales. But I have not thoroughly reviewed again this PR, and not yet Alexander's one either. I think I will do this tomorrow.